### ISSUE TITLE
Refactor UI to use functional component

### ISSUE THREAD

**Issue by tczhao** [argoproj/argo-workflows#9810]
# Summary

Say for UI features, new feature added in workflow page would likely be useful in archived workflow as well,
but right now we have both class and functional components at various parts of the UI, porting features from one part to another can be a pain.

class component file list

- [x] 1. ui/src/app/app.tsx:
- [ ] 2. ui/src/app/archived-workflows/components/archived-workflow-details/archived-workflow-details.tsx:
- [ ] 3. ui/src/app/archived-workflows/components/archived-workflow-filters/archived-workflow-filters.tsx:
- [ ] 4. ui/src/app/archived-workflows/components/archived-workflow-list/archived-workflow-list.tsx:
- [x] 5. ui/src/app/cluster-workflow-templates/components/cluster-workflow-template-list/cluster-workflow-template-list.tsx:
- [ ] 6. ui/src/app/reports/components/reports.tsx:
- [ ] 7. ui/src/app/shared/components/base-page.tsx:
- [ ] 8. ui/src/app/shared/components/checkbox-filter/checkbox-filter.tsx:
- [ ] 9. ui/src/app/shared/components/dropdown/dropdown.tsx:
- [ ] 10. ui/src/app/shared/components/error-panel.tsx:
- [ ] 11. ui/src/app/shared/components/graph/coffman-graham-sorter.ts:
- [ ] 12. ui/src/app/shared/components/graph/dfs-sorter.ts:
- [ ] 13. ui/src/app/shared/components/graph/types.ts:
- [ ] 14. ui/src/app/shared/components/inline-table/inline-table.tsx:
- [ ] 15. ui/src/app/shared/components/input-filter.tsx:
- [ ] 16. ui/src/app/shared/components/nudge.tsx:
- [ ] 17. ui/src/app/shared/components/pagination-panel.tsx:
- [ ] 18. ui/src/app/shared/components/resource-editor/resource-editor.tsx:
- [ ] 19. ui/src/app/shared/components/tags-input/tags-input.tsx:
- [ ] 20. ui/src/app/shared/conditions-panel.tsx:
- [ ] 21. ui/src/app/shared/list-watch.ts:
- [ ] 22. ui/src/app/shared/resources-duration.tsx:
- [ ] 23. ui/src/app/shared/retry-observable.ts:
- [ ] 24. ui/src/app/shared/retry-watch.ts:
- [ ] 25. ui/src/app/shared/scoped-local-storage.ts:
- [ ] 26. ui/src/app/shared/services/archived-workflows-service.ts:
- [ ] 27. ui/src/app/shared/services/cluster-workflow-template-service.ts:
- [ ] 28. ui/src/app/shared/services/cron-workflow-service.ts:
- [ ] 29. ui/src/app/shared/services/event-service.ts:
- [ ] 30. ui/src/app/shared/services/event-source-service.ts:
- [ ] 31. ui/src/app/shared/services/info-service.ts:
- [ ] 32. ui/src/app/shared/services/sensor-service.ts:
- [ ] 33. ui/src/app/shared/services/workflow-template-service.ts:
- [ ] 34. ui/src/app/shared/services/workflows-service.ts:
- [ ] 35. ui/src/app/userinfo/components/user-info.tsx:
- [ ] 36. ui/src/app/workflows/components/submit-workflow-panel.tsx:
- [ ] 37. ui/src/app/workflows/components/workflow-creator-info/workflow-creator-info.tsx:
- [ ] 38. ui/src/app/workflows/components/workflow-dag/workflow-dag-render-options-panel.tsx:
- [ ] 39. ui/src/app/workflows/components/workflow-dag/workflow-dag.tsx:
- [ ] 40. ui/src/app/workflows/components/workflow-drawer/workflow-drawer.tsx:
- [ ] 41. ui/src/app/workflows/components/workflow-filters/workflow-filters.tsx:
- [ ] 42. ui/src/app/workflows/components/workflow-labels/workflow-labels.tsx:
- [ ] 43. ui/src/app/workflows/components/workflow-panel/workflow-panel.tsx:
- [ ] 44. ui/src/app/workflows/components/workflow-timeline/workflow-timeline.tsx:
- [ ] 45. ui/src/app/workflows/components/workflow-yaml-viewer/workflow-yaml-viewer.tsx:
- [ ] 46. ui/src/app/workflows/components/workflows-list/workflows-list.tsx:
- [ ] 47. ui/src/app/workflows/components/workflows-row/workflows-row.tsx:
- [ ] 48. ui/src/app/workflows/components/workflows-toolbar/workflows-toolbar.tsx:

---

plan to refactor them in batches

---

**Message from the maintainers**:

Love this enhancement proposal? Give it a 👍. We prioritise the proposals with the most 👍.

**Comment by terrytangyuan** [comment]
There are also other good UI/UX related issues if you want to help: https://github.com/orgs/argoproj/projects/29

**Comment by tczhao** [comment]
> @tczhao Would you like to pick this up?

Sure :)

**Comment by tczhao** [comment]
> There are also other good UI/UX related issues if you want to help: https://github.com/orgs/argoproj/projects/29

Looks good

**Comment by agilgur5** [comment]
> * [ ]  8. ui/src/app/shared/components/checkbox-filter/checkbox-filter.tsx:

Completed `8.` as part of #11739

**Comment by agilgur5** [comment]
> * [ ]  37. ui/src/app/workflows/components/workflow-creator-info/workflow-creator-info.tsx:
> * [ ]  42. ui/src/app/workflows/components/workflow-labels/workflow-labels.tsx:
> * [ ]  43. ui/src/app/workflows/components/workflow-panel/workflow-panel.tsx:
> * [ ]  45. ui/src/app/workflows/components/workflow-yaml-viewer/workflow-yaml-viewer.tsx:
> * [ ]  47. ui/src/app/workflows/components/workflows-row/workflows-row.tsx:

Completed the above 5 in #11791

**Comment by agilgur5** [comment]
Some of the files listed in the opening comment are also not React components. Quick way to tell: they end in `.ts` and not `.tsx`

**Comment by agilgur5** [comment]
> * [ ]  36. ui/src/app/workflows/components/submit-workflow-panel.tsx:

* [ ] ui/src/app/workflows/components/resubmit-workflow-panel.tsx:
* [ ] ui/src/app/workflows/components/retry-workflow-panel.tsx:

@toyamagu-2021 do you think you'd be able to refactor the submit/resubmit/retry workflow panels to functional components? Since you had created the latter two, I figure you're most familiar with them 🙂 

Feel free to take any others on as well if you'd like!

**Comment by toyamagu-2021** [comment]
OK, I'll do that!

**Comment by agilgur5** [comment]
> * [x]  2. ui/src/app/archived-workflows/components/archived-workflow-details/archived-workflow-details.tsx:
> * [x]  3. ui/src/app/archived-workflows/components/archived-workflow-filters/archived-workflow-filters.tsx:
> * [x]  4. ui/src/app/archived-workflows/components/archived-workflow-list/archived-workflow-list.tsx:

These 3 components no longer exist as the Archived Workflows UI was merged with the Workflows UI in #11121
(and neither does the `archived-workflow-service`, but that was already checked off [above](https://github.com/argoproj/argo-workflows/issues/9810#issuecomment-1712651952) as it is not a React component)

* [ ] ui/src/app/shared/components/error-boundary.tsx
  * This was not listed and notably _cannot_ be converted to a functional component. Apparently there is still no way to write [error boundaries](https://react.dev/reference/react/Component#catching-rendering-errors-with-an-error-boundary) with functional components. Though wrapping with https://github.com/bvaughn/react-error-boundary is suggested as a way to remove the `class`-based code from one's own codebase

**Comment by agilgur5** [comment]
> * [x]  6. ui/src/app/reports/components/reports.tsx:
> * [ ]  7. ui/src/app/shared/components/base-page.tsx:
> * [x]  35. ui/src/app/userinfo/components/user-info.tsx:
> * [ ]  46. ui/src/app/workflows/components/workflows-list/workflows-list.tsx:

Completed `Reports` and `UserInfo` in #11793 and #11794. Partially removed `BasePage` in those as well.

Both of these used to use `BasePage`, as does `WorkflowsList`. Once `WorkflowsList` is migrated, `BasePage` can be deleted entirely (which also unblocks `react-router-dom` upgrades).

I'm working on a PR for `WorkflowsList` as well, which is one of the most complicated components in the codebase (if not the most). There are some other PRs currently open for `WorkflowsList` as well though, so would want those to be merged soon so that we have less merge conflicts to resolve.

**Comment by agilgur5** [comment]
> * [x]  18. ui/src/app/shared/components/resource-editor/resource-editor.tsx:
> * [x]  19. ui/src/app/shared/components/tags-input/tags-input.tsx:
> * [x]  40. ui/src/app/workflows/components/workflow-drawer/workflow-drawer.tsx:

Completed these 3 components in #11800

**Comment by agilgur5** [comment]
> * [ ]  9. ui/src/app/shared/components/dropdown/dropdown.tsx:
> * [ ] 15. ui/src/app/shared/components/input-filter.tsx:
> * [ ]  39. ui/src/app/workflows/components/workflow-dag/workflow-dag.tsx:
> * [ ]  44. ui/src/app/workflows/components/workflow-timeline/workflow-timeline.tsx:

Would anyone like to take some or all of these remaining 4?

We've also got one new one that was missed in #11803:
> * [ ]  53. ui/src/app/shared/components/parameters-input/parameters-input.tsx

cc @tetora1053 @juijeong8324 @yunwoo-yu  who may be interested

**Comment by juijeong8324** [comment]
> > * [ ]   9. ui/src/app/shared/components/dropdown/dropdown.tsx:
> > * [ ]  15. ui/src/app/shared/components/input-filter.tsx:
> > * [ ]   39. ui/src/app/workflows/components/workflow-dag/workflow-dag.tsx:
> > * [ ]   44. ui/src/app/workflows/components/workflow-timeline/workflow-timeline.tsx:
> > * [ ]   48. ui/src/app/workflows/components/workflows-toolbar/workflows-toolbar.tsx:
> 
> Would anyone like to take some or all of these remaining 5? (Please write which you want to take if so)
> 
> We've also got another new one that was missed in #11803 (I didn't realize it existed so didn't have it in [the initial list](https://github.com/argoproj/argo-workflows/issues/9810#issuecomment-1712664770), my bad!):
> 
> * [ ]   53. ui/src/app/shared/components/parameters-input/parameters-input.tsx
> 
> cc @tetora1053 @juijeong8324 @yunwoo-yu who may be interested 🙂

Oh I'll do that!! I'm interested in 9, 48 and 53!
* [ ]   9. ui/src/app/shared/components/dropdown/dropdown.tsx:
* [ ]  48. ui/src/app/workflows/components/workflows-toolbar/workflows-toolbar.tsx:
* [ ]  53. ui/src/app/shared/components/parameters-input/parameters-input.tsx

I will take **53** first! If anyone interested in 9 or 48, it's okay take them😀😀

**Comment by yunwoo-yu** [comment]
> > * [ ]   9. ui/src/app/shared/components/dropdown/dropdown.tsx:
> > * [ ]  15. ui/src/app/shared/components/input-filter.tsx:
> > * [ ]   39. ui/src/app/workflows/components/workflow-dag/workflow-dag.tsx:
> > * [ ]   44. ui/src/app/workflows/components/workflow-timeline/workflow-timeline.tsx:
> > * [ ]   48. ui/src/app/workflows/components/workflows-toolbar/workflows-toolbar.tsx:
> 
> Would anyone like to take some or all of these remaining 5? (Please write which you want to take if so)
> 
> We've also got another new one that was missed in #11803 (I didn't realize it existed so didn't have it in [the initial list](https://github.com/argoproj/argo-workflows/issues/9810#issuecomment-1712664770), my bad!):
> 
> * [ ]   53. ui/src/app/shared/components/parameters-input/parameters-input.tsx
> 
> cc @tetora1053 @juijeong8324 @yunwoo-yu who may be interested 🙂

Hello 😄

* [ ]   9. ui/src/app/shared/components/dropdown/dropdown.tsx:

I'm interested in all of them! But I'll start with number 9 in the list order 👍.

**Comment by agilgur5** [comment]
> * [x]   7. ui/src/app/shared/components/base-page.tsx:
> * [x]   41. ui/src/app/workflows/components/workflow-filters/workflow-filters.tsx:
> * [x]   46. ui/src/app/workflows/components/workflows-list/workflows-list.tsx:

Completed these all in #11891. That was actually a lot more challenging than I expected and is probably one of the most involved refactors I've done in several years 😮
It legit made my head hurt at times fixing bugs and getting the semantics to be equivalent 😅  (to be fair, I do have chronic migraines 🤕 )

That was _the_ most difficult refactor in the UI codebase though -- which is why I took it on myself -- so with that over, it's just the final stretch of the remaining 5. Hopefully new contributors above can tackle those effectively 😉

**Comment by tetora1053** [comment]
> > * [ ]   9. ui/src/app/shared/components/dropdown/dropdown.tsx:
> > * [ ]  15. ui/src/app/shared/components/input-filter.tsx:
> > * [ ]   39. ui/src/app/workflows/components/workflow-dag/workflow-dag.tsx:
> > * [ ]   44. ui/src/app/workflows/components/workflow-timeline/workflow-timeline.tsx:
> > * [ ]   48. ui/src/app/workflows/components/workflows-toolbar/workflows-toolbar.tsx:
> 
> Would anyone like to take some or all of these remaining 5? (Please write which you want to take if so)
> 
> We've also got another new one that was missed in #11803 (I didn't realize it existed so didn't have it in [the initial list](https://github.com/argoproj/argo-workflows/issues/9810#issuecomment-1712664770), my bad!):
> 
> * [ ]   53. ui/src/app/shared/components/parameters-input/parameters-input.tsx
> 
> cc @tetora1053 @juijeong8324 @yunwoo-yu who may be interested 🙂

ok! I take 15, 39, 44 (no one has taken those yet, right?)

**Comment by juijeong8324** [comment]
@tetora1053 yeah you are right! 
I already done 53, so I will take 48!

**Comment by tetora1053** [comment]
I'll tackle no.39 with another PR. (cause I'm afraid it might take a little longer than the other two)

**Comment by juijeong8324** [comment]
Hello, I'm sorry to be late in tackling number 48 (I had a cold and wasn't feeling well...) I'll open the PR within this week!! (==tomorrow😆) Thank you!!

**Comment by agilgur5** [comment]
no worries I was sick this week with a pretty bad cold too 😷 . hope you are feeling better!
